src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,315 @@
+            return $@"<root>
+  <Application private_binpath = ""{outputPath}""/>
+  <Assembly
+    codebase = ""{outputFileName}""
+    name = ""{fileName}""
+    version = ""0.0.0.0""
+    snapshot_id = ""1""
+    replaceable = ""True""
+  />
+</root>";
+        }
+
+        protected override Task ApplyAsync(DesignTimeInputsDelta value)
+        {
+            var applied = AppliedValue.Value;
+            var designTimeInputs = applied.Inputs.ToBuilder();
+            var designTimeSharedInputs = applied.SharedInputs.ToBuilder();
+
+            var addedDesignTimeInputs = new List<string>();
+            var removedDesignTimeInputs = new List<string>();
+
+            foreach (var item in value.AddedItems)
+            {
+                // If a property changes we might be asked to add items that already exist, so we just ignore them as we're happy using our existing CancellationSeries
+                if (!designTimeInputs.ContainsKey(item))
+                {
+                    designTimeInputs.Add(item, new CancellationSeries());
+                    addedDesignTimeInputs.Add(item);
+                }
+            }
+
+            foreach (var item in value.AddedSharedItems)
+            {
+                designTimeSharedInputs.Add(item);
+            }
+
+            foreach (var item in value.RemovedItems)
+            {
+                if (designTimeInputs.TryGetValue(item, out var series))
+                {
+                    series?.Dispose();
+                    designTimeInputs.Remove(item);
+                    removedDesignTimeInputs.Add(item);
+                }
+                designTimeSharedInputs.Remove(item);
+            }
+
+            AppliedValue = new ProjectVersionedValue<DesignTimeInputsItem>(new DesignTimeInputsItem
+            {
+                Inputs = designTimeInputs.ToImmutable(),
+                SharedInputs = designTimeSharedInputs.ToImmutable(),
+            }, value.DataSourceVersions);
+
+            // Fire off the events
+            var buildManager = BuildManager.First().Value as VSBuildManager;
+            if (value.AddedSharedItems.Count > 0)
+            {
+                // if the shared items changed then all TempPEs are dirty, because shared items are included in all TempPEs
+                foreach (var item in designTimeInputs.Keys)
+                {
+                    buildManager.OnDesignTimeOutputDirty(item);
+                }
+            }
+            else
+            {
+                // otherwise just the ones that we've added are dirty
+                foreach (var item in addedDesignTimeInputs)
+                {
+                    buildManager.OnDesignTimeOutputDirty(item);
+                }
+            }
+            foreach (var item in removedDesignTimeInputs)
+            {
+                buildManager.OnDesignTimeOutputDeleted(item);
```

Comment:
Are we on the UI thread here?

Reply:
I thought ProjectHostBridge guaranteed that (isn't that the whole point? Only changing AppliedValue on the  UI thread) but I see now that ReferencesHostBridge has a VerifyOnUIThread at the start so I will add that here.

Reply:
Anywhere we think we're on the UI thread; assert.